Misc. updates for NED

Mehrad0711 opened the pull request.

Mehrad0711 committed:
Update almond_type_mapping.json

Mehrad0711 committed:
bootleg: method refactoring

so we can reuse the new method it in some of our analysis scripts

Mehrad0711 committed:
Add scripts for analyzing NED results

Mehrad0711 committed:
Rename bootleg_post_process_types to ned_normalized_types

- the argument is not specific to bootleg anymore and can be used with Naive and Oracle too
- it now hast three  modes: no, yes, and force. previous functionlity was doing only yes and no.

Mehrad0711 committed:
Use marisa_trie for alias2qids and qid2typeqid to save memory

Mehrad0711 committed:
Remove --database_lookup_method argument

and some cleanups

Mehrad0711 committed:
Update ned scripts

Mehrad0711 committed:
Make ned code work again

Mehrad0711 committed:
Force type for certain entities

these entites are frequently disambiguated incorrectly by bootleg. we map them directly to correct type. probably need to introduce an entity-to-type mapping similar to current type-to-type mapping

Mehrad0711 committed:
server: fix bug when ned_retrieve_method is not bootleg

Mehrad0711 committed:
Add oracle model only for types

any entity that was not detected by bootleg will be ignored

Mehrad0711 committed:
Fixes to make tests pass

s-jse commented on genienlp/arguments.py:
```diff
@@ -427,7 +423,12 @@ def parse_argv(parser):
+    parser.add_argument(
+        '--ned_normalize_types',
+        type=str,
+        choices=['no', 'yes', 'force'],
```
`--ned_type_normalization` with `choices=['off', 'lenient', 'strict']` makes it more obvious what this argument does.

s-jse commented on genienlp/ned/ned_utils.py:
```diff
@@ -230,6 +231,9 @@
         'home assistant',
+        'key words',
+        'keywords',
+        'list of books',
```
Probably cleaner if this list of banned words is kept in a separate json file.

gcampax commented on genienlp/arguments.py:
```diff
@@ -401,12 +401,8 @@ def parse_argv(parser):
-        '--max_entity_len', type=int, default=4, help='Maximum length for entities when ngrams database_lookup_method is used '
-    )
+    parser.add_argument('--min_entity_len', type=int, default=1, help='Minimum token-length of entities in ngram-based lookup')
+    parser.add_argument('--max_entity_len', type=int, default=4, help='Maximum token-length of entities in ngram-based lookup')
```
I am like 50% confident that I saw this argument in Bootleg code as well, so why this help text talks about ngram-based lookup?

Mehrad0711 replied:
Yes. This argument is shared between the two scripts. Here it's used for the naive NED approach which does NER right before training (as opposed to bootleg which does it in a separate step by running run_bootleg.py)

gcampax replied:
This is confusing. How about "Minimum token-length of entities (does not apply to Bootleg)"?

gcampax commented on genienlp/arguments.py:
```diff
@@ -427,7 +423,12 @@ def parse_argv(parser):
+        type=str,
+        choices=['no', 'yes', 'force'],
+        help='Normalize types. yes: attempt to map; if unsuccessful use original. force: attempt to map; if unsuccessful drop the type.',
+    )
```
What's the equivalent of the previous option here? The config backcompat code says "yes", but I thought the previous logic was "force"

Mehrad0711 replied:
previous approach was using "yes" logic.

gcampax commented on genienlp/ned/bootleg.py:
```diff
@@ -64,10 +63,13 @@ def __init__(self, args):
             self.entityqid2typenames = ujson.load(fin)
+
+        # large print
+        self.entityqid2typenames['Q1548123'] = ['format']
```
Uh? Why is this here? In the code?

Mehrad0711 replied:
`qid2typenames` map qids to an unordered list of types. I noticed the type we want is sometimes not ranked highly. So I have to do some reordering. I don't want to modify the files directly cause I'd forget what I changed later. Instead do it in the code for visibility and version tracking.

gcampax replied:
This is the wrong way around it. Changing the code is harder. We don't want hacks in the code that confuse readers and contributors. If you want to version the bootleg input files, you can do it in the workdirs (with some light-weight override file, not by tracking the full bootleg file)

gcampax commented on genienlp/ned/bootleg.py:
```diff
@@ -135,116 +137,110 @@ def extract_mentions(self, input_path):
+            all_qids = all_qids[: self.args.max_qids_per_entity]
+            all_probs = all_probs[: self.args.max_qids_per_entity]
+
+        if not is_banned(alias):
```
```
if is_banned(alias):
    return
```

and you save one indentation level everywhere, making things more readable.

gcampax commented on genienlp/ned/scripts/analyze_bootleg_results.py:
```diff
@@ -0,0 +1,97 @@
+    all_titles = Counter()
+    all_new_titles = Counter()
+
+    fout = open('results.txt', 'w')
```
This should be a command-line argument. You cannot assume you can write to the current directory.

gcampax commented on genienlp/ned/scripts/analyze_bootleg_results.py:
```diff
@@ -0,0 +1,97 @@
+
+args = parser.parse_args()
+
+if __name__ == '__main__':
```
Why is this not a normal genienlp command?

Mehrad0711 replied:
These scripts are for local data analysis. I don't want to crowd genienlp commands with data processing/ analysis scripts.

gcampax replied:
This is a generally usable script that could be part of everyone's workflow. It's not a custom one-off script. I think it's worth making it part of genienlp proper.

Mehrad0711 replied:
My main concern is this will be an invitation for more scripts. As a general rule, I think we should exclude data manipulation scripts that do not rely on a specific model/ tokenizer and have them in genie-toolkit or workdirs. I can add these for now since you and Sina are both in favor and Kevin may wanna take a look too.

gcampax replied:
I agree, anything that's project/model specific should go into genie-workdirs.

gcampax commented on genienlp/run_bootleg.py:
```diff
@@ -96,10 +96,10 @@ def parse_argv(parser):
 
     parser.add_argument(
-        '--min_entity_len', type=int, default=1, help='Minimum length for entities when ngrams database_lookup_method is used '
+        '--min_entity_len', type=int, default=1, help='Minimum token-length of entities in ngram-based lookup of bootleg'
```
"ngram-based lookup" is implementation detail, the point is that entities shorter than that will not be retrieved

gcampax commented on genienlp/ned/bootleg.py:
```diff
@@ -135,116 +137,110 @@ def extract_mentions(self, input_path):
+                    if alias in ['houghton mifflin', 'ciudad de buenos aires']:
+                        all_types = ['award']
+                    elif alias in ['blanche', 'biblioteca breve', '3rd edition', '4th edition']:
+                        all_types = ['version, edition, or translation']
```
This is not the place for this stuff.

Also, to have this stuff at all in Bootleg is meaningless: you're just raising your accuracy on the dev set by looking at oracle results and cleaning up by hand, it's not scientific.

Mehrad0711 replied:
I'm reaching the conclusion that for certain entities we need to map them directly to the correct type.
It happens in 2 cases:
1) bootleg partly detected the entity and thus got the Qid wrong. 
2) bootleg alias2qids has incorrect entries where the book’s writer Qid is associated with the book name instead of the book itself, thus mapping to “person” type instead of “book”.

In none of those cases changing type mapping will fix the problem. Also, this mapping is only done for training set, not dev/ test set.
I added this as if-else in the code to see how experiments turn out. I'm eventually gonna move it to a json file similar to type mapping.

gcampax replied:
Yeah don't merge this hacks to the code. Add an override file if you need to, but I still don't think you need to.

gcampax commented on tests/test_NED.sh:
```diff
@@ -22,7 +22,7 @@ do
     # check if result file exists
     if test ! -f $workdir/model_$i/eval_results/valid/almond_dialogue_nlu.tsv ; then
-        echo "File not found!"
+        echo "File not found!"database_lookup_method
```
Mistake here?

gcampax reviewed (changes requested):
Btw, as a general rule, you should rebase/squash patches like "make X work again" or "fixed to make tests pass" into whatever patch broke things. Ideally, tests should pass after _every_ commit, so if there are regressions not caught by the tests we can bisect and test all commits to find the culprit.

s-jse commented on genienlp/ned/scripts/oracle_vs_bootleg.py:
```diff
@@ -0,0 +1,98 @@
+import argparse
```
Should this and `analyze_bootleg_results` be added as genienlp scripts?

Mehrad0711 replied:
These scripts are for local data analysis. I don't want to crowd genienlp commands with data processing/ analysis scripts.

Mehrad0711 committed:
Address PR comments

- change default names for add_entities_to_text and ned_normalize_types
- improve code style and comments
- move banned_phrases to a text file

Mehrad0711 committed:
Address PR comments (2)
- improve xxx_entity_len help description
- remove direct changes to bootleg database files in the code

Mehrad0711 committed:
Add ned scripts to genienlp commands

gcampax reviewed (approved):
Ok to merge when tests are happy
